### PR TITLE
[ramda] improve concat | dissoc | groupBy | toLower | toUpper | toPairs | toPairsIn

### PR DESCRIPTION
**Pull request by couzic**
## concat()
Support tuples and literal strings:
```ts
const result1: [1, 2, 3, 4] = R.concat([1, 2] as const, [3, 4] as const);
const result2: 'ABCDEF' = R.concat('ABC', 'DEF');
```

## dissoc()
Omit key:
```ts
const result: { a: 'A' } = R.dissoc({ a: 'A', b: 'B' } as const);
```

## groupBy()

As seen in tests (without changing a line in [test code](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ramda/test/groupBy-tests.ts)), the result type is more precise:

Before:
```ts
{
    [index: string]: {
        score: number;
        name: string;
    }[];
}
```

After:
```ts
Record<"F" | "D" | "C" | "B" | "A", {
    score: number;
    name: string;
}[]>
```

## toLower() / toUpper()
Support string literals:
```ts
const lower: 'test' = R.toLower('TesT');
const upper: 'TEST' = R.toUpper('TesT');
```

## toPairs() /  toPairsIn()
Return type is array of tuples:
```ts
const result: Array<['a', 1]> = R.toPairs({ a: 1 } as const)
```

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ramda ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ramda@latest/))
was missing the following properties:
1. sequence
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 87e5f22c497f32a41d64425643b9aa15822cb626
</p>

File: types/ramda/index.d.ts [hunk 1]
```diff
@@ -30,17 +30,16 @@
...
+// TypeScript Version: 4.2
 
 import * as _ from "ts-toolbelt";
 import {
-    Arity0Fn,
     Arity1Fn,
     Arity2Fn,
     AssocPartialOne,
     ComposeWithFns,
-    Dictionary,
```
Comment (Nicholaiii) [inline types/ramda/index.d.ts:43]: These changes to imports doesn't seem relevant to your PR.

File: types/ramda/test/toPairsIn-tests.ts [hunk 2]
```diff
@@ -1,13 +1,11 @@
...
-  [k: string]: string;
-  x = 'X';
-  y = 'Y';
+  x: 'X';
+  y: 'Y';
 }
 
 () => {
   const f = new F();
-  const a1 = R.toPairsIn(f); // => [['x','X'], ['y','Y']]
```
Comment (Nicholaiii) [inline types/ramda/test/toPairsIn-tests.ts:11]: Why is this test removed? Shouldn't it still work?

**Review by Nicholaiii — CHANGES_REQUESTED** [review]
I would like to see the original tests be kept in place, so we can guarantee functionality is not severed. 

I would also prefer to not have unrelated changes, like those to the imports

File: types/ramda/index.d.ts (hunk 1 above)
Comment (couzic) [reply types/ramda/index.d.ts:43]: Yes you're right, here I was just applying the boy scout rule by removing some old unused imports. I can revert that change if you want, I don't mind much either way.

Comment (Nicholaiii) [reply types/ramda/index.d.ts:43]: You're right, it **is** a good rule of thumb, I just think it deserves it's own PR, so conversation about the changes can happen where relevant. I do believe this specific import is kept in case zipObj would use it again, there's a TODO for it.

File: types/ramda/test/toPairsIn-tests.ts (hunk 2 above)
Comment (couzic) [reply types/ramda/test/toPairsIn-tests.ts:11]: Good catch !
I was unnecessarily introducing a breaking change.
I restored the old tests and fixed the issue. I did the same thing with `toPairs()`, there was the same problem.

**Review by Nicholaiii — APPROVED** [review]
👍

**Comment by couzic** [comment]
Ready to merge

File: types/ramda/index.d.ts [hunk 3]
```diff
@@ -30,8 +30,9 @@
 //                 Nicholai Nissen <https://github.com/Nicholaiii>
 //                 Mike Deverell <https://github.com/devrelm>
 //                 Jorge Santana <https://github.com/LORDBABUINO>
+//                 Mikael Couzic <https://github.com/couzic>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 3.7
+// TypeScript Version: 4.2
```
Comment (devrelm) [inline types/ramda/index.d.ts:35]: @couzic TS 4.2 hasn't been released yet. Can you put in another PR to revert this back to at least 4.1?

**Comment by xuemengfei** [comment]
got error with
```
typescript@3.9.9

ramda@0.27.1

@types/ramda@0.27.38
```

```
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(373,64): error TS1110: Type expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(373,76): error TS1005: ';' expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(375,109): error TS1110: Type expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(377,83): error TS1110: Type expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(379,79): error TS1110: Type expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(1935,118): error TS1110: Type expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(1935,134): error TS1005: ';' expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(1935,136): error TS1128: Declaration or statement expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(1935,141): error TS1109: Expression expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(1944,120): error TS1110: Type expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(1944,136): error TS1005: ';' expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(1944,138): error TS1128: Declaration or statement expected.
[tsc] node_modules/_@types_ramda@0.27.38@@types/ramda/index.d.ts(1944,143): error TS1109: Expression expected.
[tsc] npm ERR! code ELIFECYCLE
[tsc] npm ERR! errno 2
[tsc] npm ERR! project@ tsc: `ets && tsc`
```
